docs/DesignDocs/RemoteAttestationCollaterals.md

```diff
@@ -0,0 +1,292 @@
+    ///<  |-----------------------|
+    ///<  |  reserved    | type   |
+    ///<  |-----------------------|
+    ///<  |  num_elements         |
+    ///<  |-----------------------|
+    ///<  |  buffer_size          |
+    ///<  |-----------------------|
+    ///<  |                       |
+    ///<  |  array(uint32_t)      |
+    ///<  |  size of num_elements |
+    ///<  |-----------------------|
+    ///<  |  Data                 |
+    ///<  |_______________________|
+    ///<
+    uint8_t buffer[];              ///< Buffer of offsets + data
+
+} oe_collaterals_t;
+```
+
+### New Collateral functions
+
+`collaterals.c`
+```C
+/*! \struct sgx_collaterals
+ *
+ * \brief SGX collateral structure
+ *
+ * The generic oe_collaterals_t structure is parsed and converted into this
+ * internal structure.  The order of the generic data elements should
+ * coincide with the order of the fields in this structure.
+ */
+typedef struct _sgx_collaterals_t
+{
+    uint8_t* tcb_info;                      ///< TCB info
+    size_t tcb_info_size;                   ///< TCB Info size
+    uint8_t* tcb_issuer_chain;              ///< PEM format
+    size_t tcb_issuer_chain_size;           ///< Size of the tcb_issuer_chain
+    uint8_t* crl[3];                        ///< CRLs
+    size_t crl_size[3];                     ///< CRLs sizes
+    uint8_t* crl_issuer_chain[3];           ///< PEM format
+    size_t crl_issuer_chain_size[3];        ///< Size of each crl_issuer_chain
+
+    uint8_t* qe_id_info;                    ///< QE Identity info
+    size_t qe_id_info_size;                 ///< QE Identity size
+    uint8_t* qe_id_issuer_chain;            ///< PEM format
+    size_t qe_id_issuer_chain_size;         ///< Size of qe_id_issuer_chain
+
+    uint8_t* creation_datetime;             ///< Time the collaterals were generated
+    size_t create_datetime_size;            ///< The size of creation_datetime.  Should be 24.
+
+} sgx_collaterals_t;
+
+/**
+ * Returns the collaterals needed for verification of a remote OE report.
+ *
+ * This function is only available inside an enclave.
+ *
+ * @param[in] enclave The instance of the enclave that will be used.
+ * @param[in] collaterals_buffer The buffer containing the collaterals to parse.
+ * @param[in] collaterals_buffer_size The size of the **collaterals_buffer**.
+ *
+ * @retval OE_OK The collaterals were successfully retrieved.
+ */
+oe_result_t oe_get_collaterals(
+    uint8_t** collaterals_buffer,
+    size_t* collaterals_buffer_size);
+
+/**
+ * Free up any resources allocated by oe_get_collateras()
+ *
+ * @param[in] collaterals_buffer The buffer containing the collaterals.
+ */
+void oe_free_collaterals(uint8_t* collaterals_buffer);
+
+/**
+ * Verify the integrity of the report and its signature,
+ * with optional collaterals that is associated with the report. If
+ * the collaterals are not specified, this function will fetch
+ * the collaterals.  This only applies to remote reports.  For
```

Comment:
I wonder if this function should be useable for local reports as well or if `oe_verify_report()` should be extended to accept collaterals as well. I am thinking if the local enclave has an attestation policy to be used against a local report, how would that work?

Reply:
collaterals apply only to remote attestation and not local attestation.

Reply:
My impression was that collaterals would also include attestation policy. If collaterals are not to be used for local attestation, how will attestation policy be passed to local attestation?

Reply:
> collaterals apply only to remote attestation and not local attestation.

I don't think that's true for endorsements in general, especially for CyRes where there can be multiple layers, not just 2, in the attestation chain.

Reply:
@jiria 
We need to maintain backward compatible for oe_verify_report() API at least for one past public release. So changing the signature of oe_verify_report() is not part of this design plan.

At OE SDK, we don't directly support "Policy" information in our APIs, instead, we provide API channels for an involved entity to pass "blob" of data to another in the attestation stack.  In upcoming attestation plug-in model PR (with design doc, it will be posted after this one is in because it depends on collaterals APIs), it provides more customizable way for a plug-in to funnel additional data (such as policy data) to a verifier.

@dthaler

We could like to make it support both remote and local attestation. Could you share with us on the sample endorsement data in CyRes case?

Reply:
I thought oe_verify_report() was considered "experimental" and so could be broken.  But if not, then agree we should just mark oe_verify_report as deprecated, and add a new function alongside it that's not deprecated.

Regarding cyres data, JiriA can probably provide more details than I can, but concepts like PCK cert chain (e.g., where the manufacturer vounches that it manufactured a device) and CRL would apply equally to cyres.

Reply:
I updated the design with oe_get_evidence() and a oe_verify_evidence() that will support both local and remote Attestation.